README
NAME

    Dancer::Layout::Bootstrap - Boostrap layout for Dancer

SYNOPSIS

      # 1. Copy the included template files to your dancer application
      # 2. Change the "layout" parameter in config.yml to "bootstrap3"
      # 3. Dance On!
    
      # Preview the layout
      plackup -a bin\app.pl

ABSTRACT

    Template and dummy application that applies the Bootstrap CSS to
    Dancer.

SUPPORTED FEATURES

    If you want to show alerts etc., the main template already provides
    support for that via Dancer::Plugin::FlashMessage.

        flash success => sprintf "Welcome back, %s",
            session('user')->{name};

    ... will produce a nice, green "Welcome back" message for the user.

    All flash message output is HTML escaped for security reasons.

TEMPLATE FILES IN THIS DISTRIBUTION

    The meat of this package are the template and Bootstrap and jQuery
    files. These don't get installed but live in the distribution. You will
    need to copy the following files and directories into your application
    directory, at least until Dancer resp. the dancer application comes
    with a way to specify default layouts when generating a new application
    scaffold.

    You will likely want to use either Bootstrap 2 or Bootstrap 3. See the
    appropriate list of files to copy.

 LIST OF FILES TO COPY

  Bootstrap 3

        public/css/bootstrap-3/bootstrap.css Bootstrap 3
        public/css/bootstrap-3/bootstrap.min.css Bootstrap 3
        public/css/fonts/glyphicons-halflings-regular.eot Bootstrap 3
        public/css/fonts/glyphicons-halflings-regular.svg Bootstrap 3
        public/css/fonts/glyphicons-halflings-regular.ttf Bootstrap 3
        public/css/fonts/glyphicons-halflings-regular.woff Bootstrap 3
        public/css/fonts/glyphicons-halflings-regular.woff2 Bootstrap 3
        public/javascripts/bootstrap-3/bootstrap.js Bootstrap 3
        public/javascripts/bootstrap-3/bootstrap.min.js Bootstrap 3
        public/javascripts/jQuery/jquery-3.1.1.min.js jQuery 3.1.1 (for Bootstrap 3)
        public/javascripts/jQuery/jquery-ui-1.12.min.js jQuery UI (for Bootstrap 3)
        views/index.tt        Index page with Bootstrap layout
        views/layouts/bootstrap3.tt    The Bootstrap layout

  Bootstrap 2

        public/css/bootstrap-responsive.css   Bootstrap 2.3.0
        public/css/bootstrap.css      Bootstrap 2.3.0
        public/css/bootstrap.min.css  Bootstrap 2.3.0
        public/images/glyphicons-halflings-white.png  Glyphicons
        public/images/glyphicons-halflings.png        Bootstrap 2.0.3
        public/javascripts/bootstrap-2.3.0/bootstrap-alert.js Bootstrap 2.3.0
        public/javascripts/bootstrap-2.3.0/bootstrap.js       Bootstrap 2.3.0
        public/javascripts/bootstrap-2.3.0/bootstrap.min.js   Bootstrap 2.3.0
        public/javascripts/jQuery/jquery-1.7.2.min.js jQuery 1.7.2
        public/javascripts/jQuery/jquery.ui.widget.js jQuery 1.7.2
        views/index.tt        Index page with Bootstrap layout
        views/layouts/bootstrap.tt    The Bootstrap layout

NEEDED MODULES

    The simple templating included with Dancer does not properly support
    user log-in / log out, so the template needs Template::Toolkit.

SEE ALSO

    http://twitter.github.com/bootstrap/ - Bootstrap

    http://bootswatch.com/ - Themes for Bootstrap

    http://glyphicons.com/ - Icons from Glyphicons Free, licensed under CC
    BY 3.0.

    http://jquery.com/ - jQuery 1.7.2 , jQuery 3

    http://www.lavishbootstrap.com/ - create new colour schemes

REPOSITORY

    The public repository of this module is
    https://github.com/Corion/dancer-layout-bootstrap.

SUPPORT

    The public support forum of this module is https://perlmonks.org/.

BUG TRACKER

    Please report bugs in this module via the RT CPAN bug queue at
    https://rt.cpan.org/Public/Dist/Display.html?Name=Dancer-Layout-Bootstr
    ap or via mail to dancer-layout-bootstrap-Bugs@rt.cpan.org.

AUTHOR

    Max Maischein corion@cpan.org

COPYRIGHT (c)

    Copyright 2014-2017 by Max Maischein corion@cpan.org.

LICENSE

    This module is released under the same terms as Perl itself.



Changes
0.04 20170112
    . Updated Bootstrap 3 template and CSS
    . Now uses jQuery 3

0.03 20160324
    . Fixed broken link to icons
    . Added Boostrap 3 template and CSS
    . Also see L<http://getbootstrap.com/migration/> on what to change
      in your HTML templates

0.02 20130224
    . Updated to Bootstrap 2.3.0
    . Added Login template
        Watch out for Dancer::Limbo, a complete set of templates
        and logic, maybe coming soon.
    . Released on CPAN

0.01 never released on CPAN


Makefile.PL
#!perl -w
use strict;
use ExtUtils::MakeMaker qw(WriteMakefile);

my @tests = map { glob $_ } 't/*.t', 't/*/*.t';
my $module = 'Dancer::Layout::Bootstrap';
(my $main_file = "lib/$module.pm" ) =~ s!::!/!g;
(my $repo = lc $module) =~ s!::!-!g;
(my $distname = $module) =~ s!::!-!g;

# I should maybe use something like Shipwright...
regen_README($main_file);
#regen_EXAMPLES();

my %module = (
    MIN_PERL_VERSION => '5.006',
    META_MERGE => {
        resources => {
            repository => {
                web => "https://github.com/Corion/$repo",
                url => "git://github.com/Corion/$repo.git",
                type => 'git',
            },
            bugtracker => 'http://rt.cpan.org/Public/Dist/Display.html?Name=' . $distname,
            license    => 'http://dev.perl.org/licenses/',
        },
        dynamic_config   => 0, # we promise to keep META.* up-to-date
        x_static_install => 1, # we are pure Perl and don't do anything fancy
    },
    BUILD_REQUIRES => {
                        'Exporter'   => '0',
                        'Carp'       => '0',
    },
    'NAME'          => $module,
    'LICENSE'       => 'perl',
    'VERSION_FROM'  => $main_file,
    'PREREQ_PM'     => {
                       'Template' => 2, # for output to HTML
                       'Test::More' => 0,
                       'YAML'       => 0,
                       'Dancer'     => 0, # Was developped under 1.3095
                       'Dancer::Plugin::FlashMessage' => 0,
                      }, # e.g., Module::Name => 1.1
    ABSTRACT_FROM   => $main_file, # retrieve abstract from module
    AUTHOR          => 'Max Maischein <corion@cpan.org>',
    test            => { TESTS => join( ' ', @tests ) },
    clean           => { FILES => "$distname-*" },
);

sub get_module_info { %module }

if( ! caller ) {
    # I should maybe use something like Shipwright...
    regen_README($main_file);
    #regen_EXAMPLES();
    WriteMakefile1(get_module_info);
};

sub WriteMakefile1 {  #Written by Alexandr Ciornii, version 0.21. Added by eumm-upgrade.
    my %params=@_;
    my $eumm_version=$ExtUtils::MakeMaker::VERSION;
    $eumm_version=eval $eumm_version;
    die "EXTRA_META is deprecated" if exists $params{EXTRA_META};
    die "License not specified" if not exists $params{LICENSE};
    if ($params{BUILD_REQUIRES} and $eumm_version < 6.5503) {
        #EUMM 6.5502 has problems with BUILD_REQUIRES
        $params{PREREQ_PM}={ %{$params{PREREQ_PM} || {}} , %{$params{BUILD_REQUIRES}} };
        delete $params{BUILD_REQUIRES};
    }
    delete $params{CONFIGURE_REQUIRES} if $eumm_version < 6.52;
    delete $params{MIN_PERL_VERSION} if $eumm_version < 6.48;
    delete $params{META_MERGE} if $eumm_version < 6.46;
    delete $params{META_ADD} if $eumm_version < 6.46;
    delete $params{LICENSE} if $eumm_version < 6.31;
    delete $params{AUTHOR} if $] < 5.005;
    delete $params{ABSTRACT_FROM} if $] < 5.005;
    delete $params{BINARY_LOCATION} if $] < 5.005;
    
    WriteMakefile(%params);
}

sub regen_README {
    eval {
        require Pod::Readme;

        my $parser = Pod::Readme->new();

        # Read POD from Module.pm and write to README
        $parser->parse_from_file($_[0], 'README');
    };
    eval {
        require Pod::Markdown;

        my $parser = Pod::Markdown->new();

        # Read POD from Module.pm and write to README
        $parser->parse_from_file($_[0]);
        open my $fh, '>', 'README.mkdn'
            or die "Couldn't open 'README.mkdn': $!";
        print $fh $parser->as_markdown;
    };
}

sub regen_EXAMPLES {
    my $perl = $^X;
    (my $target = $main_file) =~ s!\.pm$!/Examples.pm!;
    
    if( -d 'examples' ) {
        warn "(Re)Creating $target\n";
        if ($perl =~/\s/) {
            $perl = qq{"$perl"};
        };
        my $examples = `$perl -w examples/gen_examples_pod.pl $module`;
        if ($examples) {
            $examples =~ s/\r\n/\n/g;
            if (open my $out, '>', $target) {
                binmode $out;
                print $out $examples;
            };
        };
    };
};


lib/Dancer/Layout/Bootstrap.pm
package Dancer::Layout::Bootstrap;
use Dancer ':syntax';
use Dancer::Plugin::FlashMessage;

=head1 NAME

Dancer::Layout::Bootstrap - Boostrap layout for Dancer

=head1 SYNOPSIS

  # 1. Copy the included template files to your dancer application
  # 2. Change the "layout" parameter in config.yml to "bootstrap3"
  # 3. Dance On!

  # Preview the layout
  plackup -a bin\app.pl

=head1 ABSTRACT

Template and dummy application that applies the
Bootstrap CSS to Dancer.

=head1 SUPPORTED FEATURES

If you want to show alerts etc., the main
template already provides support for that
via L<Dancer::Plugin::FlashMessage>.

    flash success => sprintf "Welcome back, %s",
        session('user')->{name};

... will produce a nice, green "Welcome back"
message for the user.

All flash message output is HTML escaped for security reasons.

=head1 TEMPLATE FILES IN THIS DISTRIBUTION

The meat of this package are the template and Bootstrap and jQuery files.
These don't get installed but live in the distribution. You will need to
copy the following files and directories into your application
directory, at least until L<Dancer> resp. the C<dancer> application comes
with a way to specify default layouts when generating a new application
scaffold.

You will likely want to use either Bootstrap 2 or Bootstrap 3. See the
appropriate list of files to copy.

=head2 LIST OF FILES TO COPY

=head3 Bootstrap 3

    public/css/bootstrap-3/bootstrap.css Bootstrap 3
    public/css/bootstrap-3/bootstrap.min.css Bootstrap 3
    public/css/fonts/glyphicons-halflings-regular.eot Bootstrap 3
    public/css/fonts/glyphicons-halflings-regular.svg Bootstrap 3
    public/css/fonts/glyphicons-halflings-regular.ttf Bootstrap 3
    public/css/fonts/glyphicons-halflings-regular.woff Bootstrap 3
    public/css/fonts/glyphicons-halflings-regular.woff2 Bootstrap 3
    public/javascripts/bootstrap-3/bootstrap.js Bootstrap 3
    public/javascripts/bootstrap-3/bootstrap.min.js Bootstrap 3
    public/javascripts/jQuery/jquery-3.1.1.min.js jQuery 3.1.1 (for Bootstrap 3)
    public/javascripts/jQuery/jquery-ui-1.12.min.js jQuery UI (for Bootstrap 3)
    views/index.tt        Index page with Bootstrap layout
    views/layouts/bootstrap3.tt    The Bootstrap layout

=head3 Bootstrap 2

    public/css/bootstrap-responsive.css   Bootstrap 2.3.0
    public/css/bootstrap.css      Bootstrap 2.3.0
    public/css/bootstrap.min.css  Bootstrap 2.3.0
    public/images/glyphicons-halflings-white.png  Glyphicons
    public/images/glyphicons-halflings.png        Bootstrap 2.0.3
    public/javascripts/bootstrap-2.3.0/bootstrap-alert.js Bootstrap 2.3.0
    public/javascripts/bootstrap-2.3.0/bootstrap.js       Bootstrap 2.3.0
    public/javascripts/bootstrap-2.3.0/bootstrap.min.js   Bootstrap 2.3.0
    public/javascripts/jQuery/jquery-1.7.2.min.js jQuery 1.7.2
    public/javascripts/jQuery/jquery.ui.widget.js jQuery 1.7.2
    views/index.tt        Index page with Bootstrap layout
    views/layouts/bootstrap.tt    The Bootstrap layout

=head1 NEEDED MODULES

The simple templating included with Dancer does not properly
support user log-in / log out, so the template needs
L<Template::Toolkit>.

=head1 SEE ALSO

L<http://twitter.github.com/bootstrap/> - Bootstrap

L<http://bootswatch.com/> - Themes for Bootstrap

L<http://glyphicons.com/> - Icons from Glyphicons Free, licensed under CC BY 3.0.

L<http://jquery.com/> - jQuery 1.7.2 , jQuery 3

L<http://www.lavishbootstrap.com/> - create new colour schemes

=cut

our $VERSION = '0.04';

get '/' => sub {
    template 'index';
};

get '/user/login' => sub {
    template 'login';
};

post '/user/login' => sub {
    session user => { name => params->{'name'} };

    flash success => sprintf "Welcome back, %s",
        session->{user}->{name};
    
    return redirect '/';
};

get '/user/logout' => sub {
    my $user= session('user')->{name};
    session user => undef;

    flash success => sprintf "Goodbye, %s",
        $user;
    
    redirect '/';
};

true;

=head1 REPOSITORY

The public repository of this module is
L<https://github.com/Corion/dancer-layout-bootstrap>.

=head1 SUPPORT

The public support forum of this module is
L<https://perlmonks.org/>.

=head1 BUG TRACKER

Please report bugs in this module via the RT CPAN bug queue at
L<https://rt.cpan.org/Public/Dist/Display.html?Name=Dancer-Layout-Bootstrap>
or via mail to L<dancer-layout-bootstrap-Bugs@rt.cpan.org>.

=head1 AUTHOR

Max Maischein C<corion@cpan.org>

=head1 COPYRIGHT (c)

Copyright 2014-2017 by Max Maischein C<corion@cpan.org>.

=head1 LICENSE

This module is released under the same terms as Perl itself.

=cut